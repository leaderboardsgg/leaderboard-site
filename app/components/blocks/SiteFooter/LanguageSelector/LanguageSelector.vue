<script setup lang="ts">
import { ref, useI18n, watch } from '#imports'
const { locale, locales, setLocale } = useI18n()

const localLocale = ref(locale.value)

watch(localLocale, () => {
  setLocale(localLocale.value)
})
</script>

<template>
  <!-- eslint-disable vuejs-accessibility/form-control-has-label -->
  <select v-model="localLocale" class="language-selector" role="listbox">
    <option
      v-for="currentLocale in locales"
      :key="currentLocale.code"
      :value="currentLocale.code"
      :selected="currentLocale.code === locale"
    >
      {{ currentLocale.name }}
    </option>
  </select>
</template>

<style lang="postcss" scoped>
.language-selector {
  @apply bg-white border border-current rounded;
  @apply flex items-center px-2 py-2 relative;
  @apply w-32;
}
.language-selector > option {
  @apply flex items-center py-2;
}
</style>
