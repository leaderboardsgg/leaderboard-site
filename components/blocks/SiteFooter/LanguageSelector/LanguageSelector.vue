<template>
  <select
    v-model="selected"
    class="language-selector"
    role="listbox"
    @change="onChange"
  >
    <option v-for="locale in locales" :key="locale.code" :value="locale.code">
      {{ locale.name }}
    </option>
  </select>
</template>

<script lang="ts">
import { defineComponent } from '@vue/composition-api';
import { Locale } from '@/types';

export default defineComponent({
  name: 'LanguageSelector',
  data: () => ({
    selected: '',
  }),
  computed: {
    currentLocale(): Locale | undefined {
      return this.locales.find(
        (locale: Locale) => locale.code === this.$i18n.locale,
      );
    },
    locales(): Locale[] {
      return this.$i18n.locales as Locale[];
    },
  },
  mounted() {
    this.selected = this.currentLocale?.code ?? '';
  },
  methods: {
    onChange(event: any): void {
      this.$i18n.setLocale(event.target.value);
      this.selected = event.target.value;
    },
  },
});
</script>

<style scoped>
.language-selector {
  @apply bg-white border border-current rounded;
  @apply flex items-center px-2 py-2 relative;
}

.language-selector > option {
  @apply flex items-center py-2;
}
</style>
