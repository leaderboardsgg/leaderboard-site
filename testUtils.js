import Vue from 'vue';
import { RouterLinkStub } from '@vue/test-utils';
import userEvent from '@testing-library/user-event';
import { render } from '@testing-library/vue';

export const NuxtLinkStub = Vue.component('NuxtLinkStub', {
  extends: RouterLinkStub,
  render(h) {
    const temp = h(
      this.tag || 'a',
      {
        attrs: {
          ...this.$vnode.data.attrs,
          ...this.$attrs,
          href: this.to,
        },
      },
      this.$slots.default,
    );

    return temp;
  },
});

export const stubbedRender = (Component, options) => {
  return render(Component, {
    stubs: { NuxtLink: NuxtLinkStub, ...options?.stubs },
    ...options,
  });
};

export * from '@testing-library/vue';
export * from '@vue/test-utils';
export { userEvent as fireEvent };
