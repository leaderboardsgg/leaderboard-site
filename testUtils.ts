import Vue, { ComponentOptions } from 'vue';
import { config, RouterLinkStub, VueClass } from '@vue/test-utils';
import userEvent from '@testing-library/user-event';
import { render, RenderOptions } from '@testing-library/vue';
import { MockI18nOptions } from '@/types';

export const NuxtLinkStub = Vue.component<
  { tag: string; to: string },
  unknown,
  unknown
>('NuxtLinkStub', {
  extends: RouterLinkStub,
  render(h) {
    const temp = h(
      this.tag || 'a',
      {
        attrs: {
          ...this.$vnode.data?.attrs,
          ...this.$attrs,
          href: this.to,
        },
      },
      this.$slots.default,
    );

    return temp;
  },
});

export const mockI18n = (options?: MockI18nOptions) => {
  const mockEnglishLocale = { code: 'en', name: 'English' };

  return {
    $i18n: {
      locales: [
        mockEnglishLocale,
        { code: 'es', name: 'Spanish' },
        { code: 'fr', name: 'French' },
        ...(options?.locales ?? []),
      ],
      locale: options?.defaultLocale ?? mockEnglishLocale,
      setLocale: options?.setLocale ?? jest.fn(),
    },
  };
};

export const stubbedRender = (
  Component: VueClass<Vue> | ComponentOptions<Vue>,
  options?: RenderOptions<Vue>,
) => {
  return render(Component, {
    mocks: {
      ...mockI18n(),
      ...options?.mocks,
    },
    stubs: {
      NuxtLink: NuxtLinkStub,
      'svg-icon': true,
      ...options?.stubs,
    },
    ...options,
  });
};

config.mocks.$t = (key: string) => key;
config.mocks.localeRoute = (route: string) => route;

export * from '@testing-library/vue';
export * from '@vue/test-utils';
export { userEvent as fireEvent };
