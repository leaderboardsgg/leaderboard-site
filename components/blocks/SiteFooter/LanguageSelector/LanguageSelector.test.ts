import LanguageSelector from './LanguageSelector.vue';
import { fireEvent, mockI18n, stubbedRender } from '@/testUtils';

const mocks = {
  ...mockI18n(),
};

describe('<LanguageSelector />', () => {
  it('should render without crashing', () => {
    const { unmount } = stubbedRender(LanguageSelector, { mocks });

    unmount();
  });

  it('renders correctly', () => {
    const { container } = stubbedRender(LanguageSelector, { mocks });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a dropdown with all the locales as options', () => {
    const { getByText } = stubbedRender(LanguageSelector, { mocks });

    expect(getByText('English')).toBeInTheDocument();
    expect(getByText('Spanish')).toBeInTheDocument();
    expect(getByText('French')).toBeInTheDocument();
  });

  describe('when an option is selected', () => {
    it('should call the $i18n.setLocale method with the selected locale', async () => {
      const { getByRole } = stubbedRender(LanguageSelector, { mocks });
      const dropdown = getByRole('listbox');

      await fireEvent.selectOptions(dropdown, 'es');

      expect(mocks.$i18n.setLocale).toHaveBeenCalledWith('es');
    });
  });
});
