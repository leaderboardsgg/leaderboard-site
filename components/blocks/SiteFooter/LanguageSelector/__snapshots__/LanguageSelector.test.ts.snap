// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LanguageSelector /> renders correctly 1`] = `
<select
  class="language-selector"
  role="listbox"
>
  <option
    value="en"
  >
    
    English
  
  </option>
  <option
    value="es"
  >
    
    Spanish
  
  </option>
  <option
    value="fr"
  >
    
    French
  
  </option>
</select>
`;
